build-tools/cmd/zz-e2e-perf.go

```diff
@@ -0,0 +1,460 @@
+		}
+	}
+
+	// Build the image
+	// It also pushes to the cache registry if needed
+	err = c.buildDockerImage(cachedImage)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Handler for the "push" sub-command
+func (c *cmdE2EPerf) pushCmd(cmd *cobra.Command, args []string) error {
+	// Target image and tag
+	destImage := c.getDestImage()
+
+	// Push the image
+	err := c.pushDockerImage(destImage)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Handler for the "build-and-push" sub-command
+func (c *cmdE2EPerf) buildAndPushCmd(cmd *cobra.Command, args []string) error {
+	// Target image and tag
+	destImage := c.getDestImage()
+
+	// Cached image and tag
+	cachedImage, err := c.getCachedImage()
+	if err != nil {
+		return err
+	}
+
+	// Try to copy the image between the cache and the target directly, without pulling first
+	// This will fail if the image is not in cache, and that's ok
+	if c.flags.CacheRegistry != "" {
+		fmt.Printf("Trying to copy image %s directly from cache %s\n", destImage, cachedImage)
+		err = crane.Copy(cachedImage, destImage)
+		if err == nil {
+			// If there's no error, we're done
+			fmt.Println("Image copied from cache directly! You're all set.")
+			return nil
+		}
+
+		// Copying the image failed, so we'll resort to build + push
+		fmt.Printf("Copying image directly from cache failed with error '%s'. Will build image.\n", err)
+	} else {
+		fmt.Println("Cache registry not set: will not use cache")
+	}
+
+	// Build the image
+	// It also pushes to the cache registry if needed
+	err = c.buildDockerImage(cachedImage)
+	if err != nil {
+		return err
+	}
+
+	// Push the image
+	err = c.pushDockerImage(destImage)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Returns the cached image and tag
+func (c *cmdE2EPerf) getCachedImage() (string, error) {
+	// Get the hash of the files in the directory
+	hashDir, err := c.getHashDir()
+	if err != nil {
+		return "", err
+	}
+
+	// If cache is enable, try pulling from cache first
+	cachedImage := fmt.Sprintf("%s/%s-%s:%s-%s", c.flags.CacheRegistry, c.cmdType, c.flags.Name, c.flags.TargetOS, hashDir)
+	return cachedImage, nil
+}
+
+// Returns the target image and tag
+func (c *cmdE2EPerf) getDestImage() string {
+	return fmt.Sprintf("%s/%s-%s:%s", c.flags.DestRegistry, c.cmdType, c.flags.Name, c.flags.DestTag)
+}
+
+// Returns the directory where the app is stored
+func (c *cmdE2EPerf) getAppDir() string {
+	if c.cmdType == "perf" {
+		return filepath.Join(c.flags.AppDir, "perf")
+	} else {
+		return c.flags.AppDir
+	}
+}
+
+// Builds a Docker image for the app
+// It also pushes it to the cache registry if that's enabled
+func (c *cmdE2EPerf) buildDockerImage(cachedImage string) error {
+	destImage := c.getDestImage()
+	appDir := c.getAppDir()
+
+	// First, check if the image has its own Dockerfile
+	dockerfile := filepath.Join(appDir, c.flags.Name, c.flags.Dockerfile)
+	_, err := os.Stat(dockerfile)
+	if err != nil {
```

Comment:
Other than pointing this at the 'shared' Dockerfile, is there a point for building here? Wont' the docker build do that right after this?

Reply:
No, apps that don't have a Dockerfile are built outside of Docker, then the Dockerfile just copies the binary in the image. This has been like this for a while, and I like that it helps reducing the number of go.mod files (and makes creating tests much simpler)